README.md
# Jour cli

> Simple, and hopefully effective, personal journaling tool

Use jour to help you create daily journal entries, right from your terminal.

## Install

```
$ npm install --global jour-cli
```

## Usage

```
Usage: jour [options]

Options:
  -r, --register <directory>                    Set jour directory
  -t, --template <template-name>                Set which template to use. Name, or path to template
  -i, --info                                    Display information about current jour directory
  -v, --version                                 Display version info
  -c, --connect [optional-git-origin-repo-url]  Initialize, and optionally connect to remote, git in current jour directory.
  -s, --save                                    Adds, and commits, all files in current jour directory to git.
  -u, --upload                                  Pushes all git changes to remote repository.
  -V, --verbose                                 Print verbose debug logging
  -o, --open                                    Open current jour directory
  -l, --locale <locale>                         Set your desired locale, for date and time formatting. eg. "en-GB" (Unicode Language Identifier)
  -a, --about                                   Display information about Jour CLI
  -d, --day <number>                            Day offset
  -w, --week <number>                           Week offset
  -y, --year <number>                           Year offset
  -ls, --list                                   List jour entries
  -lst, --list-templates                        List pre-defined templates
  -h, --help                                    display help for command

Examples:
$ jour --dir ./my/desired/jour/directory
	Set current jour directory. Jour directory is set globally.
$ jour
	Generate todays jour entry, or open existing entry if one already exists. Open in default editor.
$ jour Title for todays entry
	Generate a new daily jour entry, tagged today, with the provided written title. Unix timestamped if exists.
$ jour --template memo
	Set jour template for the current jour directory. Updates jour.json in jour directory.
$ jour --template ./my/desired/template.hbs
	Set custom jour template for the current jour directory. Updates jour.json in jour directory.

Git connection:
$ jour --connect
	Initializes git in current jour directory. Jour creates a branch named "jour" which is used by default.
$ jour --connect https://github.com/YOUR-USERNAME/my-daily-journal
	Initializes git in current jour directory, connecting to your provided remote url.
$ jour --save
	Adds, and commits, all files in current jour directory to git.
$ jour --upload
	Pushes all git changes to remote repository.

Time modifiers:
$ jour --day -1 Yesterdays journal
	Creates an entry dated today -1 day, eg. yesterday.
$ jour --week 1 In one week I will write this journal
	Creates an entry dated 1 week in the future
$ jour --year -23 Back in the day
	Creates an entry dated 23 years back in time
$ jour --day -23 --week 2 --year -3 Strange times
	Create an entry dated -23+14-365*3 = -1104 days ago.



```


> jour --register ./my

## Intended use
This tool is intended to aid you in making a daily journal. Typically, you want to configure your desired journal path once, and then use the tool to generate new entries each day from anywhere in your terminal.

There is one global jour directory at a time, so no matter which directory your terminal points to, the generated jour entry will always be put into your currently selected jour directory.

## License
MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 1.2.2 (2022-09-12)

### 1.2.1 (2021-10-30)


### Bug Fixes

* update test ([79404e1](https://github.com/stephan-nordnes-eriksen/jour/commit/79404e14c86317644b5f831f9c6d15b1196b5a11))

## 1.2.0 (2021-10-30)


### Features

* list-templates command ([7238e92](https://github.com/stephan-nordnes-eriksen/jour/commit/7238e92acedf0aa9cdd704820d596fc3295a6059))

### 1.1.4 (2021-10-28)


### Bug Fixes

* **refactor:** uppercase static methods ([75b036a](https://github.com/stephan-nordnes-eriksen/jour/commit/75b036acd4010740da85776475426b2d693f3334))

### 1.1.3 (2021-10-27)


### Bug Fixes

* replace period in filename ([f4eef76](https://github.com/stephan-nordnes-eriksen/jour/commit/f4eef76e3653b5b092a76fab5cf0e1b65cb0e7fd))

### 1.1.2 (2021-10-27)


### Bug Fixes

* save locale ([b399005](https://github.com/stephan-nordnes-eriksen/jour/commit/b3990053c5500226d7e6fc812a2e8520ed51f584))

### 1.1.1 (2021-10-27)


### Bug Fixes

* indent rules tabs ([52d2a0f](https://github.com/stephan-nordnes-eriksen/jour/commit/52d2a0f637970be87e62d15d45462a19ca532c8d))

## 1.1.0 (2021-10-23)


### Features

* add time modifiers ([3f0c6ac](https://github.com/stephan-nordnes-eriksen/jour/commit/3f0c6ace20f2c4a189ab097323697bc7ab313b9a))

### 1.0.4 (2021-10-17)


### Bug Fixes

* use correct default template ([76ad405](https://github.com/stephan-nordnes-eriksen/jour/commit/76ad405cc478e47b7353a7d41c94db7fcce8c244))

### 1.0.3 (2021-10-17)


### Bug Fixes

* adding info about issue ([140f24c](https://github.com/stephan-nordnes-eriksen/jour/commit/140f24cf45b553c1c975ca6a23474e3e4b775b86))

### 1.0.2 (2021-10-17)


### Bug Fixes

* add templates to release ([aa90cae](https://github.com/stephan-nordnes-eriksen/jour/commit/aa90cae7bdd5f91ece64d69c7441ddad3767200c))

### 1.0.1 (2021-10-17)

## [1.0.0](https://github.com/stephan-nordnes-eriksen/jour/compare/v0.0.9...v1.0.0) (2021-10-17)


### Features

* First release ([5c6d8a3](https://github.com/stephan-nordnes-eriksen/jour/commit/5c6d8a32fdaac397a159d75697472bbee46da382))

### 0.0.9 (2021-10-17)


### Bug Fixes

* set npmrc correctly ([833d82f](https://github.com/stephan-nordnes-eriksen/jour/commit/833d82f27848e87b8c8b938ab23d6b2274c2b375))

### 0.0.8 (2021-10-17)

### 0.0.7 (2021-10-17)


### Bug Fixes

* attempt to fix npmjs user ([49a37ef](https://github.com/stephan-nordnes-eriksen/jour/commit/49a37ef679d30f1d2fe9b8c3b7215e331c2a03c8))

### 0.0.6 (2021-10-17)


### Bug Fixes

* typo issue ([e02041b](https://github.com/stephan-nordnes-eriksen/jour/commit/e02041b18780e3d850f0818d0441da97c777db0e))

### 0.0.5 (2021-10-17)


### Bug Fixes

* setup github action git-account ([bee5998](https://github.com/stephan-nordnes-eriksen/jour/commit/bee5998e805bb45509895626e4b8bc176327ed9b))


src/Cli.ts
#!/usr/bin/env node
import { Command } from 'commander'
import { JourConfig } from './JourConfig.js'
import { JourError } from './JourError.js'
import { JourSystem } from './JourSystem.js'
import { Logger } from './Logger.js'


const program = new Command("jour")
//.argument('<words...>') // This seems suboptimal
program
	.option('-r, --register <directory>', 'Set jour directory')
	.option('-t, --template <template-name>', 'Set which template to use. Name, or path to template')
	.option('-i, --info', 'Display information about current jour directory')
	.option('-v, --version', 'Display version info')
	.option('-c, --connect [optional-git-origin-repo-url]', 'Initialize, and optionally connect to remote, git in current jour directory.')
	.option('-s, --save', 'Adds, and commits, all files in current jour directory to git.')
	.option('-u, --upload', 'Pushes all git changes to remote repository.')
	.option('-V, --verbose', 'Print verbose debug logging')
	.option('-o, --open', 'Open current jour directory')
	.option('-l, --locale <locale>', 'Set your desired locale, for date and time formatting. eg. "en-GB" (Unicode Language Identifier)')
	.option('-a, --about', 'Display information about Jour CLI')
	.option('-d, --day <number>', 'Day offset')
	.option('-w, --week <number>', 'Week offset')
	.option('-y, --year <number>', 'Year offset')
	.option('-ls, --list', 'List jour entries')
	.option('-lst, --list-templates', 'List pre-defined templates')
// .version("1.0.0") // TODO: Load config and set this, or maybe not?
	.addHelpText('after', `
Examples:
$ jour --dir ./my/desired/jour/directory
	Set current jour directory. Jour directory is set globally.
$ jour
	Generate todays jour entry, or open existing entry if one already exists. Open in default editor.
$ jour Title for todays entry
	Generate a new daily jour entry, tagged today, with the provided written title. Unix timestamped if exists.
$ jour --template memo
	Set jour template for the current jour directory. Updates jour.json in jour directory.
$ jour --template ./my/desired/template.hbs
	Set custom jour template for the current jour directory. Updates jour.json in jour directory.

Git connection:
$ jour --connect
	Initializes git in current jour directory. Jour creates a branch named "jour" which is used by default.
$ jour --connect https://github.com/YOUR-USERNAME/my-daily-journal
	Initializes git in current jour directory, connecting to your provided remote url.
$ jour --save
	Adds, and commits, all files in current jour directory to git.
$ jour --upload
	Pushes all git changes to remote repository.

Time modifiers:
$ jour --day -1 Yesterdays journal
	Creates an entry dated today -1 day, eg. yesterday.
$ jour --week 1 In one week I will write this journal
	Creates an entry dated 1 week in the future
$ jour --year -23 Back in the day
	Creates an entry dated 23 years back in time
$ jour --day -23 --week 2 --year -3 Strange times
	Create an entry dated -23+14-365*3 = -1104 days ago.
`)
	.action(() => {
		const programOptions = program.opts()
		Logger.SetDebugMode(!!programOptions.verbose)
		try {
			const config = JourConfig.GetJourConfig()
			const jour = new JourSystem(config)

			if (programOptions.day && ! isNaN(Number(programOptions.day))) {
				config.currentTime.setDate(config.currentTime.getDate() + Number(programOptions.day))
			}
			if (programOptions.week && ! isNaN(Number(programOptions.week))) {
				config.currentTime.setDate(config.currentTime.getDate() + 7 * Number(programOptions.week))
			}
			if (programOptions.year && ! isNaN(Number(programOptions.year))) {
				config.currentTime.setDate(config.currentTime.getDate() + 365 * Number(programOptions.year))
			}
			config.currentTime
			Logger.debug('program.args', program.args)
			switch (true) {
			case !!programOptions.register:
				Logger.debug('Running register')
				jour.Register(programOptions.dir)
				break
			case !!programOptions.template:
				Logger.debug('Running template')
				jour.Template(programOptions.template)
				break
			case !!programOptions.connect:
				Logger.debug('Running connect')
				jour.ConnectGitStorage(programOptions.connect)
				break
			case !!programOptions.save:
				Logger.debug('Running save')
				jour.Save()
				break
			case !!programOptions.upload:
				Logger.debug('Running upload')
				jour.Upload()
				break
			case !!programOptions.info:
				Logger.debug('Running info')
				jour.Info()
				break
			case !!programOptions.version:
				Logger.debug('Running version')
				jour.Version()
				break
			case !!programOptions.open:
				Logger.debug('Running open')
				jour.Open()
				break
			case !!programOptions.locale:
				Logger.debug('Running locale')
				jour.Locale(programOptions.locale)
				break
			case !!programOptions.about:
				Logger.debug('Running about')
				jour.About()
				break
			case !!programOptions.list:
				Logger.debug('Running list')
				jour.List()
				break
			case !!programOptions.listTemplates:
				Logger.debug('Running list-templates')
				jour.ListTemplates()
				break
			default:
				Logger.debug('Running default')
				Logger.debug('programOptions', programOptions)
				jour.NewJourEntry(program?.args?.join(' ') || '')
				break
			}
		} catch (error) {
			if(error instanceof JourError){
				Logger.error(error.message)
				process.exit(1)
			}
		}
	}).parse(process.argv)
